File: src/main/java/com/tomergabel/examples/java/PatternMatching3.java
package com.tomergabel.examples.java;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PatternMatching3 {

    public static void main(String[] args) {
        Pattern url = Pattern.compile("(\\w+)://([a-zA-Z0-9.-]+?)(/[a-zA-Z0-9.-/]*)?");
        Matcher matcher = url.matcher(args[0]);
        if (matcher.matches()) {
            String scheme = matcher.group(1);
            String host = matcher.group(2);
            String path = matcher.group(3);

            switch (scheme) {
                case "http":
                    System.out.println("HTTP URL detected\nHost: " + host + "\nPath: " + path);
                    break;

                case "https":
                    System.out.println("SSL-enabled HTTP URL detected\nHost: " + host + "\nPath: " + path);
                    break;

                case "ftp":
                    System.out.println("FTP URL detected\nHost: " + host + "\nPath: " + path);
                    break;

                default:
                    System.out.println("Unrecognized scheme " + scheme + " for URL " + args[0]);
                    break;
            }
        } else
            System.out.println("Not a recognized URL: " + args[0]);
    }
}

File: src/main/scala/com/tomergabel/examples/patterns/PatternMatching3.scala
package com.tomergabel.examples.patterns

/**
  * Created by tomerga on 04/04/2016.
  */
object PatternMatching3 extends App {

  val Url = "(\\w+)://([a-zA-Z0-9.-]+?)(/[a-zA-Z0-9.-/]*)?".r

  object Http {
    def unapply(url: String): Option[(String, String, Boolean)] = url match {
      case Url("http", host, path) => Some(host, path, false)
      case Url("https", host, path) => Some(host, path, true)
      case _ => None
    }
  }

  object Ftp {
    def unapply(url: String) = url match {
      case Url("ftp", host, path) => Some(host, path)
      case _ => None
    }
  }

  args.head match {
    case Http(host, path, false) => println(s"HTTP URL detected\nHost: $host\nPath: $path")
    case Http(host, path, true) => println(s"SSL-enabled HTTP URL detected\nHost: $host\nPath: $path")
    case Ftp(host, path) => println(s"FTP URL detected\nHost: $host\nPath: $path")
    case arg @ Url(scheme, _, _) => println(s"Unsupported scheme $scheme for URL: $arg")
    case arg => println(s"Not a recognized URL: $arg")
  }
}
